privval/signer_client_test.go

```diff
@@ -0,0 +1,245 @@
+package privval
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/tendermint/tendermint/libs/common"
+	"github.com/tendermint/tendermint/types"
+)
+
+type signerTestCase struct {
+	chainID       string
+	mockPV        types.PrivValidator
+	signer        *SignerClient
+	signerService *SignerDialerEndpoint
+}
+
+func getSignerTestCases(t *testing.T) []signerTestCase {
+	testCases := make([]signerTestCase, 0)
+
+	for _, dtc := range getDialerTestCases(t) {
+		chainID := common.RandStr(12)
+		mockPV := types.NewMockPV()
+
+		ve, se := getMockEndpoints(t, chainID, mockPV, dtc.addr, dtc.dialer)
+		sr, err := NewSignerClient(ve)
+		assert.NoError(t, err)
+
+		tc := signerTestCase{
+			chainID:       chainID,
+			mockPV:        mockPV,
+			signer:        sr,
+			signerService: se,
+		}
+
+		testCases = append(testCases, tc)
+	}
+
+	return testCases
+}
+
+func TestSignerClose(t *testing.T) {
+	for _, tc := range getSignerTestCases(t) {
+		err := tc.signer.Close()
+		assert.NoError(t, err)
+
+		err = tc.signerService.Stop()
+		assert.NoError(t, err)
+	}
+}
+
+func TestSignerGetPubKey(t *testing.T) {
+	for _, tc := range getSignerTestCases(t) {
+		func() {
+			defer tc.signerService.OnStop()
+			defer tc.signer.Close()
+
+			pubKey := tc.signer.GetPubKey()
+			expectedPubKey := tc.mockPV.GetPubKey()
+
+			assert.Equal(t, expectedPubKey, pubKey)
+
+			addr := tc.signer.GetPubKey().Address()
+			expectedAddr := tc.mockPV.GetPubKey().Address()
+
+			assert.Equal(t, expectedAddr, addr)
+		}()
+	}
+}
+
+func TestSignerProposal(t *testing.T) {
+	for _, tc := range getSignerTestCases(t) {
+		func() {
+			ts := time.Now()
+			want := &types.Proposal{Timestamp: ts}
+			have := &types.Proposal{Timestamp: ts}
+
+			defer tc.signerService.OnStop()
```

Comment:
usually we don't call `OnStop` directly, but instead call Service'es `Stop` func.